### PR TITLE
8287107: CgroupSubsystemFactory.setCgroupV2Path asserts with freezer controller

### PR DESCRIPTION
**Pull request by iklam**
This PR fixes a bug found on an Ubuntu host that's mostly running with cgroupv2, but there's a controller (freezer) that is mounted in cgroupv1 mode.

The container support code in the VM and JDK checks if we have simultaneously mounted v1 and v2 containers. If so, we revert to "hybrid" mode (which is essentially v1).

However, the "hybrid checks" only considers the following controllers that Java cares about:

- cpu
- cpuacct
- cpuset
- blkio
- memory
- pids

If only "freezer" is mounted in v1 mode, Java will start in V2 mode. Later, when `setCgroupV2Path()` sees the first line in /proc/self/cgroup, it runs into the assert.

```
$ cat /proc/self/cgroup
1:freezer:/
0::/user.slice/user-1001.slice/session-85.scope
```

The fix is simple -- when we get to `setCgroupV2Path()`, we have already decided that the system has not mounted any v1 controllers that we care about, so we should just ignore all the v1 controllers (which has a non-empty name).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287107](https://bugs.openjdk.java.net/browse/JDK-8287107): CgroupSubsystemFactory.setCgroupV2Path asserts with freezer controller

### Reviewers
 * [Mikhailo Seledtsov](https://openjdk.java.net/census#mseledtsov) (@mseledts - Committer) ⚠️ Review applies to [c4d8354d](https://git.openjdk.java.net/jdk/pull/8858/files/c4d8354d2abfbb96098d4d1bf462f4715d26f2fe)
 * [Severin Gehwolf](https://openjdk.java.net/census#sgehwolf) (@jerboaa - **Reviewer**) ⚠️ Review applies to [1f17b6e8](https://git.openjdk.java.net/jdk/pull/8858/files/1f17b6e85c7eaef3c7f0ea368ece3268c0d14e6a)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8858/head:pull/8858` \
`$ git checkout pull/8858`

Update a local copy of the PR: \
`$ git checkout pull/8858` \
`$ git pull https://git.openjdk.java.net/jdk pull/8858/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8858`

View PR using the GUI difftool: \
`$ git pr show -t 8858`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8858.diff">https://git.openjdk.java.net/jdk/pull/8858.diff</a>

</details>

### CODE REVIEW

**Comment by mseledts** [comment]
Change looks good to me. Thanks for the fix.

**Review by mseledts — APPROVED** [review]

File: src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java [hunk 1]
```diff
@@ -149,6 +149,10 @@ public static Optional<CgroupTypeResult> determineType(String mountInfo,
             case MEMORY_CTRL:   infos.put(MEMORY_CTRL, info); break;
             case BLKIO_CTRL:    infos.put(BLKIO_CTRL, info); break;
             case PIDS_CTRL:     infos.put(PIDS_CTRL, info); break;
+            default:
+                // There are some controllers (such as freezer) that Java doesn't
+                // care about. Just ignore them. These are not considered in the
+                // anyCgroupsV1Controller/anyCgroupsV1Controller checks.
```
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java:155]: It's not clear why this `default` is necessary. Could we just add the comment like so:

```
// Intentional fall-through. There are controllers (such as freezer) that
// Java doesn't care about. Just ignore them. Only listed controllers are
// considered in the anyCgroupsV1Controller/anyCgroupsV2Controller checks.
```

File: src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java [hunk 2]
```diff
@@ -220,6 +224,11 @@ public static Optional<CgroupTypeResult> determineType(String mountInfo,
      */
     private static void setCgroupV2Path(Map<String, CgroupInfo> infos,
                                         String[] tokens) {
+        String name = tokens[1];
+        if (!name.equals("")) {
+            // This is probably a v1 controller that we have ignored (e.g., freezer)
```
Comment (jerboaa) [inline src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java:229]: Could we add an assertion that the controller isn't in the `infos` map?
```
           if (!name.equals("")) {
            // This must be a v1 controller that we have ignored (e.g., freezer)
            assert infos.get(name) == null;
            return;
         }
```

Comment (iklam) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java]: Fixed.

File: src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java (hunk 1 above)
Comment (jerboaa) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java:155]: My bad. How about `Intentional incomplete switch. There are ...`? Anyway, why is the empty `default` case needed other than for the comment?

Comment (iklam) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java:155]: To be, the `default:` switch is a clear indication that "everything else comes here". So you won't be confused whether the comment is related to the last line just above the comment.

Comment (jerboaa) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java:155]: It confused me, fwiw. Anyway up to you. It's not super important.

Comment (jerboaa) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java]: works for me. +1. Note the typo `anyCgroupsV1Controller/anyCgroupsV2Controller` not **V1** twice.

Comment (iklam) [reply src/java.base/linux/classes/jdk/internal/platform/CgroupSubsystemFactory.java]: Oops, I'll fixed that. Thanks!

**Comment by iklam** [comment]
Thanks @jerboaa and @mseledts for the review.
/integrate

**Pull request closed without merging**